tests/raftstore/test_split_region.rs

```diff
@@ -101,28 +107,92 @@ fn test_base_split_region<T: Simulator>(cluster: &mut Cluster<T>, right_derive:
 }
 
 #[test]
 fn test_node_base_split_region_right_derive() {
     let count = 5;
     let mut cluster = new_node_cluster(0, count);
-    test_base_split_region(&mut cluster, true);
+    test_base_split_region(&mut cluster, Cluster::must_split, true);
 }
 
 #[test]
 fn test_server_base_split_region_left_derive() {
     let count = 5;
     let mut cluster = new_server_cluster(0, count);
-    test_base_split_region(&mut cluster, false);
+    test_base_split_region(&mut cluster, Cluster::must_split, false);
 }
 
 #[test]
 fn test_server_base_split_region_right_derive() {
     let count = 5;
     let mut cluster = new_server_cluster(0, count);
-    test_base_split_region(&mut cluster, true);
+    test_base_split_region(&mut cluster, Cluster::must_split, true);
+}
+
+#[test]
+fn test_node_manual_split_region_right_derive() {
+    let count = 5;
+    let mut cluster = new_node_cluster(0, count);
+    test_base_split_region(&mut cluster, Cluster::must_manual_split, false);
+}
+
+#[test]
+fn test_node_manual_split_region_left_derive() {
+    let count = 5;
+    let mut cluster = new_node_cluster(0, count);
+    test_base_split_region(&mut cluster, Cluster::must_manual_split, true);
+}
+
+#[test]
+fn test_server_manual_split_region_right_derive() {
+    let count = 5;
+    let mut cluster = new_server_cluster(0, count);
+    test_base_split_region(&mut cluster, Cluster::must_manual_split, false);
+}
+
+#[test]
+fn test_server_manual_split_region_left_derive() {
+    let count = 5;
+    let mut cluster = new_server_cluster(0, count);
+    test_base_split_region(&mut cluster, Cluster::must_manual_split, true);
+}
+
+#[test]
+fn test_server_manual_split_region_twice() {
+    let count = 5;
+    let mut cluster = new_server_cluster(0, count);
+    cluster.run();
+    let pd_client = cluster.pd_client.clone();
+
+    let (split_key, left_key, right_key) = (b"k22", b"k11", b"k33");
+    cluster.must_put(left_key, b"v1");
+    cluster.must_put(right_key, b"v3");
+
+    // Left and right key must be in same region before split.
+    let region = pd_client.get_region(left_key).unwrap();
+    let region2 = pd_client.get_region(right_key).unwrap();
+    assert_eq!(region.get_id(), region2.get_id());
+
+    let key = split_key.to_vec();
+    let c = Box::new(move |mut resp: RaftCmdResponse| {
+        let admin_resp = resp.mut_admin_response();
+        let split_resp = admin_resp.mut_split();
+        let left = split_resp.take_left();
+        let right = split_resp.take_right();
+        assert_eq!(left.get_end_key(), key.as_slice());
+        assert_eq!(region2.get_start_key(), left.get_start_key());
+        assert_eq!(left.get_end_key(), right.get_start_key());
+        assert_eq!(region2.get_end_key(), right.get_end_key());
+    });
+    cluster.split_region_by_key(&region, split_key, Some(c));
+    let c = Box::new(move |resp: RaftCmdResponse| {
+        assert!(resp.has_header());
+        assert!(resp.get_header().has_error());
```

Comment:
what is the error? stale epoch?

Reply:
Now, it is `invalid split request: ...`, originates at apply.rs#L862.

https://github.com/pingcap/tikv/blob/61eb812de8df59a23a210df2776125af8ad5b6b0/src/raftstore/store/worker/apply.rs#L862